internal/pkg/cli/deploy/env.go

```diff
@@ -33,6 +35,10 @@ type environmentDeployer interface {
 	UpdateAndRenderEnvironment(out termprogress.FileWriter, env *deploy.CreateEnvironmentInput, opts ...cloudformation.StackOption) error
 }
 
+type prefixListGetter interface {
+	CloudFrontManagedPrefixListID() (*string, error)
```

Comment:
should the API be ?
```suggestion
	CloudFrontManagedPrefixListID() (string, error)
```
And return an `ErrPrefixListIDNotFound` in case it's not there instead of of `nil` for a string.

Reply:
The reason I don't think I can do this, at least with how I've currently arranged it, is that the function `cfManagedPrefixListId` should return nil when the ingress isn't enabled for cloudfront, as that function is what handles checking if it's enabled. If I make `cfManagedPrefixListId` return a `string`, then it can't handle the situation where it's not enabled, and as such shouldn't return any id, but also shouldn't error out the creation of the `CreateEnvironmentInput` struct. This change could be possible then if I moved the checks to the function `prefixLists` but I feel that isn't a good solution as over time it causes bloat to the function.

Reply:
As a follow up TLDR, to do this I'll need to move the validation to check if the cf prefix list is needed to the function `prefixLists`. I personally prefer the check to be in `cfManagedPrefixListId` because it removes potential bloat from the more general function as more potential prefix lists can be used, but I'm happy to move that check over if you think it'd be better.

Reply:
Looking at the current implementation of CloudFrontManagedPrefixList in ec2.go, I wasn't able to see a case where it returned `nil` for the string without there being an error condition, and it does in fact return an error for when the prefix list ID isn't there - though not as a separate error type. `cfManagedPrefixListId` could always make it a `*string` if needed separate from this function?

https://github.com/aws/copilot-cli/blob/a57ecbd9203bf3e51eee195e39d60a85e9ddb86f/internal/pkg/aws/ec2/ec2.go#L462-L483